device/common/include/traccc/device/thread_id.hpp

```diff
@@ -0,0 +1,86 @@
+/**
+ * traccc library, part of the ACTS project (R&D line)
+ *
+ * (c) 2024 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+#include "traccc/definitions/qualifiers.hpp"
+
+namespace traccc::device {
+struct thread_id1 {
+#if defined(__CUDACC__)
+    TRACCC_DEVICE thread_id1() {}
+#elif defined(CL_SYCL_LANGUAGE_VERSION) || defined(SYCL_LANGUAGE_VERSION)
+    TRACCC_DEVICE thread_id1(const ::sycl::nd_item<1>& item) : m_item(item) {}
+#else
+    TRACCC_DEVICE thread_id1(unsigned int threadIdx, unsigned int blockIdx,
+                             unsigned int blockDimx, unsigned int gridDimx)
+        : m_threadIdx(threadIdx),
+          m_blockIdx(blockIdx),
+          m_blockDimx(blockDimx),
+          m_gridDimx(gridDimx) {}
+#endif
```

Comment:
I don't like this design. :frowning:

I'm on board with introducing a simple, project-specific struct that we would pass this info along with. But let's just make that a struct with a couple of integer members, and that's it. Let's not make the struct differ with SYCL and with CUDA. :thinking:

For simplicity we can make helper functions in `traccc::cuda`, `traccc:sycl`, etc. for constructing an object of that simple struct, based on the (intrinsic) variables available in CUDA, SYCL, etc. But the struct itself should just be a trivial one.

Reply:
I disagree with you on this one. Storing the variables for all platforms requires more registers on CUDA platforms, where we could just use the (much more efficient) special registers. This kind of design decreases register pressure and has the potential to increase performance!

Reply:
You cannot argue that this would be good code here. :thinking: I understand your point, but this cannot be the solution.

In such a case we'd need to do something similar to what we do with barriers. That would at least result in reasonably clean / maintainable code. Though I'm not happy about that design for this use-case either to be honest.

Whatever the case, we are not putting such a macro-hell into `traccc::device_common`.

Reply:
Yeah trust me I don't like having a pile of unhygienic C++ macros like this, but the alternative is to have more templates like for the barrier...